### DISCUSSION TITLE
Java-like enums in GDScript

### DISCUSSION THREAD

**Discussion by hydroper** [godotengine/godot-proposals#7044] [Scripting]
I was trying to create an enum whose variants are unique references instantiated from the enum type itself. Variants share similiar methods and virtual properties. E.g., instead of:

```gdscript
enum E { X, Y }
```

I'd do:

```gdscript
class_name E
const X: E = E.new(15)
const Y: E = E.new(64)

var _v: int

func _init(v: int):
    self._v = v

func value_of() -> int:
    return self._v
```

There are currently two problems with that:

- GDScript currently doesn't allow a non-constant expression to be assigned to a `const`
- The `const` definitions are only visible within the class itself.

So I need to be able to both export the `const` definitions so they're visible to the public and assign non-constant expressions to these `const`s. Are there existing proposals for this?

---

`static var`, finally introduced in Godot 4.1, instead of `const` works well:

```gdscript
class_name E

static var X: E = E.new("x")

var _s: String

func _init(s: String):
    self._s = s

func to_string() -> String:
    return self._s
```

You can use it like:

```gdscript
var s = E.X.to_string()
```

**Comment by AThousandShips** [comment]
Have you tried static for the visibility part?

**Reply by AThousandShips** [reply]
No static variables

**Reply by hydroper** [reply]
The docs says:
> `static`: Defines a static function. Static member variables are not allowed.

**Reply by theraot** [reply]
`static var` are available in Godot 4.1

**Reply by hydroper** [reply]
Is there a 4.1 release somewhere? Maybe a GitHub action built for Windows?

**Reply by theraot** [reply]
The beta is just out at the time of writing: https://godotengine.org/article/dev-snapshot-godot-4-1-beta-1/

**Reply by hydroper** [reply]
I'm gonna try it; maybe the docs should be updated to say static members are now allowed?

**Reply by AThousandShips** [reply]
Go ahead and open an issue if one isn't already present [here](https://github.com/godotengine/godot-docs/issues)

**Reply by hydroper** [reply]
@theraot Yeah, using `static var` works. It's not read-only though, I suppose; but that is really good for me!

**Comment by theraot** [comment]
If you really need `const` you can do it.

- The `const`s are visible outside of the class (if they aren't for you, it is likely due to parsing errors, because of the expression not being constant).

- You need a constant expression, and there is a way, although, you might not like it: Make the class a `Resource`, create resource files of it, then `preload` them `const blahblah:ClasssName = preload("res://path/to/resource/file.tres")` is valid.

---

The way I see it the issue is that calling `new` with constant values is not a constant expression. Perhaps is it worth changing that for a future release?

**Comment by AThousandShips** [comment]
* See also #6926
